File: containers/splunk/splunk__connectors/parquet/python/src/sender.py
import asyncio
from datetime import datetime
from typing import List, Tuple, Dict
import aiohttp
from prometheus_client import Counter
import structlog
from tenacity import retry, stop_after_attempt, wait_exponential

logger = structlog.get_logger()

EVENTS_SENT = Counter("parquet_connector_events_sent_total", "Total events sent to Splunk", ["destination"])
SEND_ERRORS = Counter("parquet_connector_send_errors_total", "Total send errors to Splunk", ["destination"])

async def send_to_splunk(
    queue: asyncio.Queue,
    splunk_url: str,
    splunk_token: str,
    batch_size: int,
    buffer_timeout: float,
):
    async with aiohttp.ClientSession() as session:
        batch = []
        last_flush = datetime.now()

        while True:
            try:
                events, file_name, count = await asyncio.wait_for(queue.get(), timeout=buffer_timeout)
                logger.info(f"Received {count} events from file {file_name}")
                batch.extend(events)

                if len(batch) >= batch_size or (datetime.now() - last_flush).total_seconds() > buffer_timeout:
                    if batch:
                        await _send_batch(session, splunk_url, splunk_token, batch)
                        batch = []
                        last_flush = datetime.now()
            except asyncio.TimeoutError:
                if batch:
                    await _send_batch(session, splunk_url, splunk_token, batch)
                    batch = []
                    last_flush = datetime.now()
            except Exception as e:
                SEND_ERRORS.labels("splunk").inc()
                logger.error(f"Send error: {e}", exc_info=True)
            await asyncio.sleep(0.1)

@retry(stop=stop_after_attempt(3), wait=wait_exponential(multiplier=1, min=0.1, max=5))
async def _send_batch(session: aiohttp.ClientSession, splunk_url: str, splunk_token: str, batch: List[Dict]):
    headers = {"Authorization": f"Splunk {splunk_token}", "Content-Type": "application/json"}
    async with session.post(splunk_url, headers=headers, json=batch) as response:
        if response.status != 200:
            SEND_ERRORS.labels("splunk").inc()
            logger.error(f"Splunk send failed: {response.status}", exc_info=True)
            raise Exception(f"HTTP {response.status}")
        logger.info(f"Sent {len(batch)} events to Splunk")
        EVENTS_SENT.labels("splunk").inc_by(len(batch))

File: containers/splunk/splunk__connectors/parquet/rust/src/sender.rs
use anyhow::{Context, Result};
use log::{error, info};
use prometheus::{Counter, Registry};
use reqwest::Client;
use serde_json::Value;
use std::sync::Arc;
use std::time::{Duration, SystemTime};
use tokio::sync::mpsc;
use tokio::time;

lazy_static::lazy_static! {
    static ref EVENTS_SENT: Counter = Counter::with_opts(prometheus::opts!(
        "parquet_connector_events_sent_total",
        "Total events sent to Splunk"
    ))
    .unwrap();
    static ref SEND_ERRORS: Counter = Counter::with_opts(prometheus::opts!(
        "parquet_connector_send_errors_total",
        "Total send errors to Splunk"
    ))
    .unwrap();
}

pub async fn send_to_splunk(
    mut rx: mpsc::Receiver<(Vec<Value>, String, usize)>,
    splunk_url: String,
    splunk_token: String,
    batch_size: usize,
    buffer_timeout: f64,
    registry: Arc<Registry>,
) -> Result<()> {
    registry.register(Box::new(EVENTS_SENT.clone()))?;
    registry.register(Box::new(SEND_ERRORS.clone()))?;

    let client = Client::new();
    let mut batch = Vec::new();
    let mut last_flush = SystemTime::now();

    loop {
        let timeout = Duration::from_secs_f64(buffer_timeout);
        let result = time::timeout(timeout, rx.recv()).await;

        match result {
            Ok(Some((events, file_name, count))) => {
                info!("Received {} events from file {}", count, file_name);
                batch.extend(events);

                if batch.len() >= batch_size
                    || last_flush.elapsed().map(|d| d.as_secs_f64() > buffer_timeout).unwrap_or(true)
                {
                    if !batch.is_empty() {
                        send_batch(&client, &splunk_url, &splunk_token, &batch).await?;
                        batch.clear();
                        last_flush = SystemTime::now();
                    }
                }
            }
            Ok(None) => break,
            Err(_) => {
                if !batch.is_empty() {
                    send_batch(&client, &splunk_url, &splunk_token, &batch).await?;
                    batch.clear();
                    last_flush = SystemTime::now();
                }
            }
        }
        tokio::time::sleep(Duration::from_millis(100)).await;
    }
    Ok(())
}

async fn send_batch(client: &Client, splunk_url: &str, splunk_token: &str, batch: &[Value]) -> Result<()> {
    let headers = [
        ("Authorization", format!("Splunk {}", splunk_token)),
        ("Content-Type", "application/json".to_string()),
    ];
    let response = client
        .post(splunk_url)
        .headers(headers.into_iter().collect())
        .json(batch)
        .send()
        .await
        .context("Failed to send to Splunk")?;

    if response.status().is_success() {
        info!("Sent {} events to Splunk", batch.len());
        EVENTS_SENT.with_label_values(&["splunk"]).inc_by(batch.len() as f64);
        Ok(())
    } else {
        SEND_ERRORS.with_label_values(&["splunk"]).inc();
        Err(anyhow::anyhow!("Splunk send failed: {}", response.status()))
    }
}